rustls/src/crypto/hmac.rs

```diff
@@ -0,0 +1,58 @@
+/// A concrete HMAC implementation, reified with a hash function.
+///
+/// You should have one object that implements this trait for HMAC-SHA256, another
+/// for HMAC-SHA384, etc.
+pub trait Hmac: Send + Sync {
+    /// Prepare to use `key` as a HMAC key.
+    fn open_key(&self, key: &[u8]) -> Box<dyn Key>;
+
+    /// Give the length of the underlying hash function.  In RFC2104 terminology this is `L`.
+    fn hash_output_len(&self) -> usize;
+}
+
+/// Maximum supported HMAC tag size: supports up to SHA512.
+pub(crate) const HMAC_MAX_TAG: usize = 64;
+
+/// A HMAC tag, stored as a value.
+#[derive(Clone)]
+pub struct Tag {
```

Comment:
Yes, that's what I meant. Why do you want to avoid monomorphization? It seems to me that code bloat should be limited assuming most downstream users will only use one `CryptoProvider` implementation.

Reply:
The most natural thing to do would be to have a `Tag` type that differs per `trait Hmac`, because each one has a different size (eg, HMAC-SHA256 needs to store 32 bytes in its tag, HMAC-SHA384 needs 48, etc.). But doing that means the associated type is on `trait Hmac` and not `trait CryptoProvider`?

Putting the `Tag` type on `trait CryptoProvider` means each provider can only support cipher suites with a single `Tag` type, and can't mix (eg) hardware-accelerated and software-only implementations unless they can all agree on what a hmac tag is.

Reply:
I'm not sure how that would work. Because `Hmac` is always used at the moment through dynamic dispatch (fundamentally because the one to use is only learned at runtime, once a cipher suite is agreed). Changing that to static dispatch would I guess mean making `SupportedCipherSuite` generic over it, but that would make cipher suites with different HMACs and hashes different types, rather than different values?

Reply:
Right, that makes sense.